service/history/workflowExecutionContext.go

```diff
@@ -1142,3 +1193,64 @@ func (c *workflowExecutionContextImpl) resetWorkflowExecution(
 func (c *workflowExecutionContextImpl) getQueryRegistry() QueryRegistry {
 	return c.queryRegistry
 }
+
+func (c *workflowExecutionContextImpl) reapplyEvents(
+	ctx context.Context,
+	events *persistence.WorkflowEvents,
+) error {
+
+	reapplyEvents := []*workflow.HistoryEvent{}
+	// TODO: need to implement Reapply policy
+	for _, event := range events.Events {
+		switch event.GetEventType() {
+		case workflow.EventTypeWorkflowExecutionSignaled:
+			reapplyEvents = append(reapplyEvents, event)
+		}
+	}
+	// there is no event to reapply
+	if len(reapplyEvents) == 0 {
+		return nil
+	}
+	domainID := events.DomainID
+	// Reapply events only reapply to the current run.
+	// Leave the run id empty will reapply events to the current run.
+	execution := &workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr(events.WorkflowID),
+	}
+	domainCache := c.shard.GetDomainCache()
+	clientBean := c.shard.GetService().GetClientBean()
+	serializer := c.shard.GetService().GetPayloadSerializer()
+	domainEntry, err := domainCache.GetDomainByID(domainID)
+	if err != nil {
+		return err
+	}
+
+	activeCluster := domainEntry.GetReplicationConfig().ActiveClusterName
+	if activeCluster == c.shard.GetClusterMetadata().GetCurrentClusterName() {
```

Comment:
please also add a TODO here, since events reapplication to self is going to have issue

Reply:
done